tests/testsuite/warning_override.rs

```diff
@@ -0,0 +1,100 @@
+//! Tests for overriding warning behavior using `build.warnings` config option.
+
+use std::sync::LazyLock;
+
+use cargo_test_support::{cargo_test, project, str, Project};
+use snapbox::data::Inline;
+
+const ALLOW_CLEAN: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+[CHECKING] foo v0.0.1 ([ROOT]/foo)
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+const ALLOW_CACHED: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+static WARN: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+const DENY: LazyLock<Inline> = LazyLock::new(|| {
+    str![[r#"
+...
+[WARNING] unused variable: `x`
+...
+[WARNING] `foo` (bin "foo") generated 1 warning
+[FINISHED] `dev` profile [unoptimized + debuginfo] target(s) in [ELAPSED]s
+
+"#]]
+});
+
+fn make_project(main_src: &str) -> Project {
```

Comment:
nit: I'm always a bit cautious of sharing code too much in tests.  In this case, I think it works as the intent of the tests isn't obscure. The main question is the `make_project` call, it did take me a second when reading the code to think "yeah, `let x = 3;`  is being meant for `main`.  One option would be to move the warning into this and call it `make_project_with_rustc_warning` or something. 

Hmm, thats a good point.  When we add the checking of cargo warnings, we'll need tests for those as well.  The names are fairly general in this code and it will all need to be modified in the follow up.  Clarifying things now can make it clearer for now and make it easier to introduce new tests in the future.

Reply:
Renamed to `make_project_with_rustc_warning`

Reply:
This change was made in a later commit.  That commit should only be updating the test for feature changes and not refactoring it.  Could you move those changes to this commit?